jscomp/ml/typecore.ml

```diff
@@ -307,7 +307,9 @@ let extract_concrete_record env ty =
 
 let extract_concrete_variant env ty =
   match extract_concrete_typedecl env ty with
-    (p0, p, {type_kind=Type_variant cstrs}) -> (p0, p, cstrs)
+    (p0, p, {type_kind=Type_variant cstrs}) 
+      when not (Ast_uncurried.typeIsUncurriedFun ty)  
+      -> (p0, p, cstrs)
```

Comment:
This was the first issue, leading to the initial bad super error being displayed. The uncurried function type would extract as a concrete variant since it matched this case when it should not.

Reply:
I'm sure you've read and know more about this, but instead of throwing an exception with Not_found if uncurried, what about an implementation that extracts the uncurried case? Something like this.
```ml
if Ast_uncurried_utils.typeIsUncurriedFun ty then
   extract_concrete_variant env (Ast_uncurried.typeExtractUncurriedFun ty)
else (p0, p, cstrs)
```

Reply:
Hey @mununki thanks for the comment! Yes I think that feel a lot nicer except that in practice it would make no difference to this function since if it is an uncurried function the extracted type would not be Type_variant? So happy to add the refactor to make it more future proof but otherwise it should never hit a case where it is an uncurried function and it can extract a concrete variant

Reply:
Sorry for my misunderstanding, if so why do we need that when clause by then?

Reply:
> Sorry for my misunderstanding, if so why do we need that when clause by then?

It's because an uncurried function is represented by this pattern

```ml
Tconstr (Pident {name = "function$"}, [tArg; _], _)
```

where tArg is the underlying type of `Tarrow`

So when trying to extract the concrete type as a variant, it matches the case of Tconstr, but is not actually variant it's essentially just a wrapper around the underlying function. So we get a false positive.